### PR TITLE
[SPARK-34141][SQL] Remove side effect from ExtractGenerator

### PR DESCRIPTION
**Pull request by tanelk**
### What changes were proposed in this pull request?

Rewrote one `ExtractGenerator` case such that it would not rely on a side effect of the flatmap function.

### Why are the changes needed?

With the dataframe api it is possible to have a lazy sequence as the `output` of a `LogicalPlan`. When exploding a column on this dataframe using the `withColumn("newName", explode(col("name")))` method, the `ExtractGenerator` does not extract the generator and `CheckAnalysis` would throw an exception.

### Does this PR introduce _any_ user-facing change?

Bugfix
Before this, the work around was to put `.select("*")` before the explode.

### How was this patch tested?

UT

### CODE REVIEW

**Comment by tanelk** [comment]
@HyukjinKwon , any chance that this bugfix could get to 3.1.1?

**Comment by tanelk** [comment]
@LuciferYang, you have been working with 2.13. 
Any recommendations on how to get a lazy Seq on both 2.12 and 2.13.
On 2.13 the Seq.view does not return a Seq anymore.

**Comment by LuciferYang** [comment]
@tanelk In Scala 2.13, `Views`  are collections whose transformation operations are non strict and they no longer extend their corresponding collection type, for example, an `IndexedSeqView` no longer extends `IndexedSeq`.

Why do we need a non strict collection here? Can it be a strict collection? 
If it must be a non strict collection, may need to define the type of `output` as `scala.collection.SeqView` for version compatibility.

**Comment by tanelk** [comment]
Well that is the test case I'm trying to cover.
in 2.12 the input can be lazy Seq, but the current code breaks then.
Perhaps it's enough to say that I manually tested it and not add the UT.

**Comment by LuciferYang** [comment]
cc @srowen

**Review by srowen — APPROVED** [review]

**Comment by LuciferYang** [comment]
> Well that is the test case I'm trying to cover.
in 2.12 the input can be lazy Seq, but the current code breaks then.
Perhaps it's enough to say that I manually tested it and not add the UT.

New UT in this pr will break the compilation of Scala 2.13,  @tanelk want to get a lazy `Seq` on both 2.12 and 2.13, but `SeqView` is not `Seq` in Scala 2.13,  do you have any suggestions ? @srowen

**Comment by srowen** [comment]
Hm why does Seq vs SeqView matter here? what's the compile error?

**Comment by LuciferYang** [comment]
image

**Comment by tanelk** [comment]
@srowen 
There is a correctness issue, when an lazy `Seq` is given as the project list (for example `SeqView`).
I did fix it, but the unit test does not compile on 2.13, because there `SeqView` does not implement `Seq` any more.

Would it be okay do not add the UT or is there some lazy `Seq` present on both 2.12 and 2.13.

**Comment by srowen** [comment]
Hm, tough call. Can you invoke the second constructor by adding `:_*` to the end of the argument? But then I suppose that's not testing what you want to test either. Can you pass a Scala Stream here maybe? not sure if that works. It may work to create a new test-suite in a 2.12-only src/test directory if that's not too much trouble. It sounds like this can't happen in 2.13 anyway so the test isn't relevant.

**Comment by tanelk** [comment]
I went ahead and added version dependent tests. Thanks for the advice.

File: sql/catalyst/src/test/scala-2.13/org/apache/spark/sql/catalyst/analysis/ExtractGeneratorSuite.scala [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+package org.apache.spark.sql.catalyst.analysis
+
+import scala.collection.immutable.LazyList
+
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.plans.logical._
+import org.apache.spark.sql.types._
+
+class ExtractGeneratorSuite extends AnalysisTest {
```
Comment (srowen) [inline sql/catalyst/src/test/scala-2.13/org/apache/spark/sql/catalyst/analysis/ExtractGeneratorSuite.scala:26]: I'd add comments in each test file noting that there is a parallel one in the other source tree, so people realize that both need to change.

Comment (srowen) [reply sql/catalyst/src/test/scala-2.13/org/apache/spark/sql/catalyst/analysis/ExtractGeneratorSuite.scala:26]: If you'll add the comments I'll merge it @tanelk

File: sql/catalyst/src/test/scala-2.12/org/apache/spark/sql/catalyst/analysis/ExtractGeneratorSuite.scala [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+class ExtractGeneratorSuite extends AnalysisTest {
+
+  test("SPARK-34141: ExtractGenerator with lazy project list") {
+    val b = AttributeReference("b", ArrayType(StringType))()
+
+    val columns = AttributeReference("a", StringType)() :: b :: Nil
+    val explode = Alias(Explode(b), "c")()
+
+    // view is a lazy seq
+    val rel = LocalRelation(output = columns.view)
```
Comment (maropu) [inline sql/catalyst/src/test/scala-2.12/org/apache/spark/sql/catalyst/analysis/ExtractGeneratorSuite.scala:36]: Is it possible for end users to get stuck on this issue? If possible, could you add end-2-end tests, too?

Comment (tanelk) [reply sql/catalyst/src/test/scala-2.12/org/apache/spark/sql/catalyst/analysis/ExtractGeneratorSuite.scala:36]: I hit this issue while using spark in java. The step before was an join, where I used `JavaConverters.collectionAsScalaIterable(Arrays.asList(columns)).toSeq()` as the join condition. The JavaConverters helper returns an lazy collection.
I can take a look at on how to e2e test this.

Comment (maropu) [reply sql/catalyst/src/test/scala-2.12/org/apache/spark/sql/catalyst/analysis/ExtractGeneratorSuite.scala:36]: Thanks!

**Comment by srowen** [comment]
Merged to master. I'm satisfied with the motivation and test.

**Comment by HyukjinKwon** [comment]
Oh, just saw it. LGTM

**Pull request closed without merging**